src/main/java/com/mathworks/ci/TestResultsViewAction.java

```diff
@@ -0,0 +1,265 @@
+
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+
+import hudson.FilePath;
+import hudson.model.Run;
+
+import jenkins.model.RunAction2;
+
+public class TestResultsViewAction implements RunAction2 {
+    private transient Run<?, ?> build;
+    private FilePath workspace;
+    private String actionID;
+    private int totalCount;
+    private int passedCount;
+    private int failedCount;
+    private int incompleteCount;
+    private int notRunCount;
+
+    public TestResultsViewAction(Run<?, ?> build, FilePath workspace, String actionID) throws InterruptedException, IOException {
+        this.build = build;
+        this.workspace = workspace;
+        this.actionID = actionID;
+        
+        totalCount = 0;
+        passedCount = 0;
+        failedCount = 0;
+        incompleteCount = 0;
+        notRunCount = 0;
+
+         try{
+            // Set test results counts
+             getTestResults();
+         } catch (InterruptedException | IOException e) {
+             e.printStackTrace();
+             throw e;
+         } catch (ParseException e) {
+             e.printStackTrace();
+         }
+    }
+
+    public List<List<TestFile>> getTestResults() throws ParseException, InterruptedException, IOException {
+        List<List<TestFile>> testResults = new ArrayList<>();
+        FilePath fl = new FilePath(new File(build.getRootDir().getAbsolutePath() + File.separator + MatlabBuilderConstants.TEST_RESULTS_VIEW_ARTIFACT + this.actionID + ".json"));
+        try (InputStreamReader reader = new InputStreamReader(Files.newInputStream(Paths.get(fl.toURI())), StandardCharsets.UTF_8)) {
+            totalCount = 0;
+            passedCount = 0;
+            failedCount = 0;
+            incompleteCount = 0;
+            notRunCount = totalCount;
+
+            JSONArray testArtifact = (JSONArray) new JSONParser().parse(reader);
+            Iterator<JSONArray> testArtifactIterator = testArtifact.iterator();
+
+            while(testArtifactIterator.hasNext()){
+                Object jsonTestSessionResults = testArtifactIterator.next();
+
+                List<TestFile> testSessionResults = new ArrayList<>();
+                Map<String, TestFile> map = new HashMap<>();
+
+                if(jsonTestSessionResults instanceof JSONArray){
+                    JSONArray jsonTestSessionResultsArray = (JSONArray) jsonTestSessionResults;
+                    Iterator<JSONObject> testSessionResultsIterator = jsonTestSessionResultsArray.iterator();
+
+                    while(testSessionResultsIterator.hasNext()){
+                        JSONObject jsonTestCase = testSessionResultsIterator.next();
+                        getTestSessionResults(testSessionResults, jsonTestCase, map);
+                    }
+                }
+                else if(jsonTestSessionResults instanceof JSONObject) {

```

Comment:
I understand it's not super intuitive. If you have a look at the [artifact](https://github.com/mathworks/jenkins-matlab-plugin/pull/360/files#diff-0454a99a615b34671023fdd246f8a948e833dad50f22f2143a90108e4de507de) we use for our unit tests, you'll see that inside the top level `JSONArray`, we have a `JSONArray` followed by a `JSONObject`. This mainly happens because of how we handle multiple test sessions' result and json encode them. Only in the case where a test session has a single test which ran, we'll have an instance of `JSONObject`. Hope this helps, feel free to reach out otherwise!